replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/ScheduledReplicatorDeleter.java

```diff
@@ -0,0 +1,248 @@
+                .getItemsForConfig("", startIndex, DEFAULT_PAGE_SIZE)
+                .stream()
+                .filter(item -> !replicatorConfigIds.contains(item.getConfigurationId()))
+                .map(ReplicationItem::getMetacardId)
+                .collect(Collectors.toSet());
+
+        if (!configlessItemIds.isEmpty()) {
+          Set<String> idsInTheCatalog = metacards.getIdsOfMetacardsInCatalog(configlessItemIds);
+          Set<String> orphanedItemIds = Sets.difference(configlessItemIds, idsInTheCatalog);
+          orphanedItemIds.forEach(replicationItemManager::deleteItem);
+
+          // todo: do we really increment the index this way?
+          startIndex += configlessItemIds.size();
+        }
+      } catch (PersistenceException e) {
+        LOGGER.debug("Failed to retrieve replication items. Continuing", e);
+        startIndex += DEFAULT_PAGE_SIZE;
+      }
+    } while (!configlessItemIds.isEmpty());
+  }
+
+  private void cleanupDeletedConfigs(List<ReplicatorConfig> replicatorConfigs) {
+    List<ReplicatorConfig> deletedConfigs =
+        replicatorConfigs.stream().filter(ReplicatorConfig::isDeleted).collect(Collectors.toList());
+
+    for (ReplicatorConfig config : deletedConfigs) {
+      final String configId = config.getId();
+      final String configName = config.getName();
+
+      if (config.deleteData()) {
+        try {
+          deleteMetacards(configId);
+        } catch (PersistenceException e) {
+          LOGGER.debug(
+              "Failed to retrieve replication items for config: {}. Deletion will be retried next poll interval.",
+              configName,
+              e);
+          return;
+        } catch (SourceUnavailableException e) {
+          LOGGER.debug(
+              "Failed to delete metacards replicated by config: {}. Deletion will be retried next poll interval.",
+              configName,
+              e);
+          return;
+        }
+      }
+
+      try {
+        replicationItemManager.deleteItemsForConfig(configId);
+      } catch (PersistenceException e) {
+        LOGGER.debug(
+            "Failed to delete replication items for config: {}. Deletion will be retried next poll interval",
+            configName,
+            e);
+        return;
+      }
+
+      try {
+        deleteReplicatorHistory(configId);
+      } catch (ReplicationPersistenceException e) {
+        LOGGER.debug(
+            "History for replicator configuration {} could not be deleted. Deletion will be retried next polling interval.",
+            configName,
+            e);
+        return;
+      }
+
+      replicatorConfigManager.remove(configId);
+    }
+  }
+
+  private void deleteMetacards(String configId)
+      throws PersistenceException, SourceUnavailableException {
+    int startIndex = 0;
+    Set<String> idsToDelete;
+
+    do {
+      idsToDelete =
+          replicationItemManager
+              .getItemsForConfig(configId, startIndex, DEFAULT_PAGE_SIZE)
```

Comment:
❓ Are we guaranteed that the start index stays consistent from one query to the next? What if new items appears in the mean time? Wouldn't that change the order and all?

Reply:
I had questions about the start index too. I did not investigate it too much yet because if something is missed, it will be caught next poll interval.